ITE/10/README.adoc

```diff
@@ -0,0 +1,288 @@
+
+In general, each unit policy applies one constraint. For example, a policy might
+mirror in-toto's artifact rules, ensuring that all the materials and products of
+some step in the supply chain are as expected. Alternatively, a policy can apply
+a property-oriented constraint to an artifact or step, such as validating the
+Provenance of an artifact or that its build process was performed in an isolated
+and hermetic build environment. Note that all of these types of policies can
+coexist and successful verification is contingent on all policy validations
+succeeding.
+
+This indicates that developers need a great deal of flexibility in specifying
+policies _and_ that any domain-specific language chosen or created to express
+policies must have the necessary capabilities to fit these requirements.
+
+Prior literature has explored the use of logic based languages and their
+derivatives for specifying and validating policies. Policies, at their core, are
+logic programs that are applied against a set of claims, expressed here as
+in-toto attestations. This ITE uses languages such as Rego and CUE as examples
+of such languages, but leaves the specific choices to the in-toto community and
+adopters. The new layout schema proposed above includes a `<POLICY_LANGUAGE>`
+attribute for each policy to enable this variety in choice.
+
+=== Validating Policies
+
+The in-toto verification workflow is updated to support different types of
+inspections as defined in this ITE. For new style inspection blocks, the
+verification workflow uses the `policyLanguage` field to decide which policy
+engine to use. If `run` is specified, the workflow executes the command or
+script as it does now.
+
+[[motivation]]
+== Motivation
+
+The original in-toto specification defined only one type of metadata that was to
+be captured during supply chain operations. As such, it had a uniform
+verification workflow and provided supply chain owners the semantics, i.e.
+artifact rules, to verify the flow of artifacts in their supply chains. ITE-6
+expanded the metadata formats to support the generation of custom attestations
+that have context-specific schemas. Current in-toto layouts as defined in the
+in-toto specification cannot be used to verify the information contained in
+these attestations without executing external scripts or commands, and cannot
+directly reason about the sources of claims made in them.
+
+[[reasoning]]
+== Reasoning
+
+The intent of ITE-6 and the in-toto Attestation framework is to enable software
+producers to emit a wide range of contextual information pertaining to their
+supply chain. The intent of this ITE is to extend ITE-6 with policy definitions
+such that attestation producers can validate the information in ITE-6
+attestations. It is clear from ITE-6 that there is not a one-size-fits-all
+solution to policy definitions. Therefore, to support a wide variety of use
+cases, this ITE proposes the use of existing policy languages such as Rego and
```

Comment:
One of my biggest reservations is that often these policy languages don't have cryptographic libraries baked into the language yet AFAIK.

Reply:
That's an interesting thought I hadn't considered. My thought process with this addition was that the users could specify constraints that would apply to claims in attestations. Would policies need cryptographic capabilities? Could you help me out with an example?

Reply:
I think the point is that "@adityasaky said 'the artifact with sha256 is cool'" is hard to express in Rego.

That's solvable in two ways: first, extend your policy language to support cryptography (e.g., make it an *authorization logic*). Second, you can just do the signature validation before passing data to the policy language. So the "facts" available to Rego would all be qualified with a "so-and-so says"

Reply:
The verification workflow essentially goes with the second option. Signatures and such are verified prior to policies being applied to claims in attestations. I'll look through this document again tomorrow to see if I can make that clearer.